File: src/Spline/ParametricEquation.java
package Spline;

import Coords.*;
import Interfaces.*;

public class ParametricEquation implements Spline {
    Equation x,y;

    public ParametricEquation(Equation x, Equation y) {
        this.x=x;
        this.y=y;
    }

    public String toString() { return "((" + x.toString() + "),(" + y.toString() + "))"; }
    public double getDerivative(double t){
        return y.getDerivative(t)/x.getDerivative(t);
    }
    public double getDDerivative(double t){
        return y.getDDerivative(t)/x.getDDerivative(t);
    }
    public double getDDDerivative(double t) { return y.getDDDerivative(t)/x.getDDDerivative(t); }
    public Point getPoint(double t) {
        return new Point(x.getPoint(t).y, y.getPoint(t).y);
    }

    public double getCurvature(double t) {
        return ((x.getDerivative(t) * y.getDDerivative(t)) - (y.getDerivative(t) * x.getDDerivative(t)))
                /(Math.pow((x.getDerivative(t) * x.getDerivative(t) + (y.getDerivative(t) * y.getDerivative(t))), 1.5));
    }
}

File: src/Spline/ParametricEquation.kt
package Spline

import Coords.*
import Interfaces.*

class ParametricEquation(private val x: Equation, private val y: Equation) {

    override fun toString(): String {
        return "(($x),($y))"
    }

    fun getDerivative(t: Double): Point {
        return Point(x.getDerivative(t), y.getDerivative(t))
    }

    fun getDDerivative(t: Double): Point {
        return Point(x.getDDerivative(t), y.getDDerivative(t))
    }

    fun getDDDerivative(t: Double): Point {
        return Point(x.getDDDerivative(t), y.getDDDerivative(t))
    }

    fun getPoint(t: Double): Point {
        return Point(x.getPoint(t).y, y.getPoint(t).y)
    }

    fun getWaypoint(t: Double): Waypoint {
        return Waypoint(x.getPoint(t).y, y.getPoint(t).y, x.getDerivative(t), y.getDerivative(t), x.getDDerivative(t), y.getDDerivative(t))
    }

    fun getCurvature(t: Double): Double {
        return (x.getDerivative(t) * y.getDDerivative(t) - y.getDerivative(t) * x.getDDerivative(t)) / Math.pow(Math.pow(x.getDerivative(t), 2.0) + Math.pow(y.getDerivative(t), 2.0), 1.5)
    }

    fun getArcLength(sample: EAccuracy, t0: Double = 0.0, t1: Double = 1.0) : Double {
        var sum = 0.0
        var i = t0
        while (i <= t1) {
            sum += Math.sqrt(Math.pow(x.getDerivative(i), 2.0) + Math.pow(y.getDerivative(i), 2.0)) * (1.0 / sample.size)
            i += 1.0 / sample.size
        }
        return sum
    }
}
